sdk/python/lib/pulumi/provider/server.py

```diff
@@ -92,36 +101,40 @@ async def Construct(self, request: proto.ConstructRequest, context) -> proto.Con
         return response
 
     @staticmethod
-    def _construct_inputs(request: proto.ConstructRequest) -> Dict[str, pulumi.Output]:
+    async def _construct_inputs(request: proto.ConstructRequest) -> Dict[str, pulumi.Input[Any]]:
 
-        def deps(key: str) -> Set[Resource]:
-            return set(DependencyResource(urn) for urn in
+        def deps(key: str) -> Set[str]:
+            return set(urn for urn in
                        request.inputDependencies.get(
                            key,
                            proto.ConstructRequest.PropertyDependencies()
                        ).urns)
 
         return {
-            k: ProviderServicer._construct_output(the_input, deps=deps(k))
+            k: await ProviderServicer._create_output(the_input, deps=deps(k))
             for k, the_input in
             rpc.deserialize_properties(request.inputs, keep_unknowns=True).items()
         }
 
     @staticmethod
-    def _construct_output(the_input: Any, deps: Set[Resource]) -> Any:
+    async def _create_output(the_input: Any, deps: Set[str]) -> Any:
+        is_resource_reference = (known_types.is_resource(the_input)
```

Comment:
This condition implies roughly this:

```
r: Resource
_create_output(r, deps=[r.urn])
```

So it depends only on its own `urn`?

It's probably fine but its getting tricky, perhaps a separate def to detect if this?